src/Microsoft.DotNet.ProjectModel/WorkspaceContext.cs

```diff
@@ -0,0 +1,427 @@
+        /// are added to the WorkspaceContext.
+        /// 
+        /// If the given path points to a project.json, all the projects it referenced as well as itself
+        /// are added to the WorkspaceContext.
+        /// </summary>
+        public static WorkspaceContext CreateFrom(string projectPath, string configuration)
+        {
+            var projectPaths = ResolveProjectPath(projectPath);
+            if (projectPaths == null || !projectPaths.Any())
+            {
+                return null;
+            }
+
+            var context = new WorkspaceContext(projectPaths, configuration);
+            return context;
+        }
+
+        public static WorkspaceContext CreateFrom(string projectPath)
+        {
+            return CreateFrom(projectPath, "Debug");
+        }
+
+        public void AddProject(string path)
+        {
+            var projectPath = NormalizeProjectPath(path);
+
+            if (projectPath != null)
+            {
+                _needRefresh = _projects.Add(path);
+            }
+        }
+
+        public void RemoveProject(string path)
+        {
+            _needRefresh = _projects.Remove(path);
+        }
+
+        public IReadOnlyList<string> GetAllProjects()
+        {
+            Refresh();
+            return _projects.ToList().AsReadOnly();
+        }
+
+        /// <summary>
+        /// Refresh the WorkspaceContext to update projects collection
+        /// </summary>
+        public void Refresh()
+        {
+            if (!_needRefresh)
+            {
+                return;
+            }
+
+            var basePaths = new List<string>(_projects);
+            _projects.Clear();
+
+            foreach (var projectDirectory in basePaths)
+            {
+                var project = GetProject(projectDirectory);
+                if (project == null)
+                {
+                    continue;
+                }
+
+                _projects.Add(project.ProjectDirectory);
+
+                foreach (var projectContext in GetProjectContexts(project.ProjectDirectory))
+                {
+                    foreach (var reference in GetProjectReferences(projectContext))
+                    {
+                        var referencedProject = GetProject(reference.Path);
+                        if (referencedProject != null)
+                        {
+                            _projects.Add(referencedProject.ProjectDirectory);
+                        }
+                    }
+                }
+            }
+
+            _needRefresh = false;
+        }
+
+        public IReadOnlyList<ProjectContext> GetProjectContexts(string projectPath)
+        {
+            return _projectContextsCache.AddOrUpdate(
+                projectPath,
+                key => AddProjectContextEntry(key, null),
+                (key, oldEntry) => AddProjectContextEntry(key, oldEntry)).ProjectContexts;
+        }
+
+        private Project GetProject(string projectDirectory)
```

Comment:
This is available from the `ProjectContext`. I'd like to see us remove these and instead have an abstraction over a "collection of ProjectContexts relating to a single project.json file", but that can happen after the PR I guess.